allennlp/semparse/contexts/atis_tables.py

```diff
@@ -65,31 +120,49 @@ def get_date_from_utterance(tokenized_utterance: List[Token],
         return datetime(year, month, day)
     return None
 
-def get_numbers_from_utterance(utterance: str) -> List[str]:
+def get_numbers_from_utterance(utterance: str, tokenized_utterance: List[Token]) -> Dict[str, List[int]]:
     """
-    Given an utterance, find all the numbers that are in the action space.
+    Given an utterance, this function finds all the numbers that are in the action space. Since we need to
+    keep track of linking scores, we represent the numbers as a dictionary, where the keys are the string
+    representation of the number and the values are lists of the token indices that triggers that number.
     """
-    numbers = []
-    numbers.extend(re.findall(r'\d+', utterance))
-    numbers.extend([str(int(num_str.rstrip('pm')) * HOUR_TO_TWENTY_FOUR + TWELVE_TO_TWENTY_FOUR)
-                    for num_str in re.findall(r'\d+', utterance)])
-
-    numbers.extend(get_times_from_utterance(utterance))
-
-    words = utterance.split(' ')
-    for word in words:
-        if word in MONTH_NUMBERS:
-            numbers.append(str(MONTH_NUMBERS[word]))
-        if word in DAY_NUMBERS:
-            numbers.append(str(DAY_NUMBERS[word]))
-        if word in MISC_TIME_TRIGGERS:
-            numbers.extend(MISC_TIME_TRIGGERS[word])
-
-    for tens, digits in zip(words, words[1:]):
-        day = ' '.join([tens, digits])
-        if day in DAY_NUMBERS:
-            numbers.append(str(DAY_NUMBERS[day]))
-    return sorted(numbers, reverse=True)
+
+    # When we use a regex to find numbers or strings, we need a mapping from
+    # the character to which token triggered it.
+    char_offset_to_token_index = {token.idx : token_idx for token_idx, token
+                                  in enumerate(tokenized_utterance)}
+
+    # We want to look up later for each time whether it appears after a word
+    # such as "about" or "approximately".
+    indices_approximate_words = [idx for idx, token in enumerate(tokenized_utterance) \
+            if token.text in APPROX_WORDS]
+    number_linking_dict = get_regex_match(r'\d+',
```

Comment:
Oh, I see, you're using this for more than just times.  Hmm...

It seems like this logic is doing two different things: (1) it is getting numbers like "Flights from 2 airlines", or something; (2) it is getting times like "flights around 7", without an "am" or "pm".  The trouble is it's not doing (2) correctly, because "flights at 7" will map to "7", not "700".  It looks like you're also double counting "7" and "7am" here, as this one will also match "7am".

So, I'd recommend treating the first case here, and you can do that with something much simpler than a regex: just iterate over the tokens in the utterance, and try `int(token.text)`, keeping the value if it succeeds.  Then leave the time matching to `get_times_from_utterance`, being sure to have this logic not match things like "7am".

Does that sound right?